File: 0278-first-bad-version/0278-first-bad-version.js
/**
 * Definition for isBadVersion()
 * 
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
var solution = function(isBadVersion) {
    /**
     * @param {integer} n Total versions
     * @return {integer} The first bad version
     */
    return function(n) {
               let s = 0, e = n - 1;
        let m;
        
        while (s <= e) {
            m = Math.floor(s + (e - s) / 2);
            
            if (isBadVersion(m)) {
                e = m - 1; // Move to the left half if it's a bad version
            } else {
                s = m + 1; // Move to the right half if it's not a bad version
            }
        }
        
        return s; // The start index will be at the first bad version 
    };
};

File: 0278-first-bad-version/0278-first-bad-version.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

// class Solution {
// public:
//     int bad=1;
//     int i;
//     int firstBadVersion(int n) {
//         i=floor(n/2);
//         if(isBadVersion(i) != isBadVersion(i+1)){
//             bad = n+1;
//         }
//         else if(isBadVersion(i) == false) {
//             firstBadVersion(n+1);
//         }else {
//             firstBadVersion(i);
//         }
//         return bad;
//     }
// };
class Solution {
public:
    int firstBadVersion(int n) {
        int s=0, e=n-1 ;
        int m = s + (e-s)/2 ;
        while (s<=e){
            if (isBadVersion(m)) e = m-1 ;
            else s = m + 1 ;
            m = s + (e-s)/2 ;
        }
        return m ;
    }
};